File: python/src/entities/obstacle.py
"""Obstacle entity"""

import random
from typing import Any
from config import OBSTACLE_SYMBOLS, NUM_LANES


class Obstacle:
    """
    Obstacle class representing hazards on the track
    Demonstrates: Type hints, default parameters, class methods
    """

    def __init__(self, lane: int, position: float, type: str = 'static') -> None:
        """
        Initialize an obstacle

        Args:
            lane: Lane where obstacle is placed (0-3)
            position: Vertical position on track
            type: Type of obstacle ('static' or 'moving')
        """
        self.lane: int = lane
        self.position: float = position
        self.type: str = type
        self.symbol: str = random.choice(OBSTACLE_SYMBOLS)
        self.speed: float = random.random() * 0.5 if type == 'moving' else 0.0

    def update(self) -> None:
        """Update obstacle position (for moving obstacles)"""
        if self.type == 'moving':
            self.position += self.speed

    def is_at(self, position: float, lane: int) -> bool:
        """
        Check if obstacle is at a specific position and lane

        Args:
            position: Position to check
            lane: Lane to check

        Returns:
            True if obstacle is at the specified location
        """
        return self.lane == lane and abs(self.position - position) < 1

    def is_off_track(self, track_length: int) -> bool:
        """
        Check if obstacle is off the track

        Args:
            track_length: Length of the track

        Returns:
            True if obstacle is beyond track bounds
        """
        return self.position < 0 or self.position > track_length

    def to_dict(self) -> dict[str, Any]:
        """
        Serialize obstacle to dictionary

        Returns:
            Serialized obstacle data
        """
        return {
            'lane': self.lane,
            'position': self.position,
            'type': self.type,
            'symbol': self.symbol
        }

    @classmethod
    def create_random(cls, position: float = 0.0) -> 'Obstacle':
        """
        Create a random obstacle

        Args:
            position: Starting position (usually 0 for top of track)

        Returns:
            New obstacle instance
        """
        lane = random.randint(0, NUM_LANES - 1)
        type_ = 'moving' if random.random() < 0.2 else 'static'  # 20% chance of moving
        return cls(lane, position, type_)

File: javascript/src/entities/obstacle.js
// Obstacle entity
import { OBSTACLE_SYMBOLS, NUM_LANES } from '../config.js';

/**
 * Obstacle class representing hazards on the track
 * Demonstrates: ES6 classes, destructuring, default parameters
 */
export class Obstacle {
  /**
   * @param {number} lane - Lane where obstacle is placed (0-3)
   * @param {number} position - Vertical position on track
   * @param {string} type - Type of obstacle ('static' or 'moving')
   */
  constructor(lane, position, type = 'static') {
    this.lane = lane;
    this.position = position;
    this.type = type;
    this.symbol = OBSTACLE_SYMBOLS[Math.floor(Math.random() * OBSTACLE_SYMBOLS.length)];
    this.speed = type === 'moving' ? Math.random() * 0.5 : 0;
  }

  /**
   * Update obstacle position (for moving obstacles)
   */
  update() {
    if (this.type === 'moving') {
      this.position += this.speed;
    }
  }

  /**
   * Check if obstacle is at a specific position and lane
   * @param {number} position - Position to check
   * @param {number} lane - Lane to check
   * @returns {boolean} - True if obstacle is at the specified location
   */
  isAt(position, lane) {
    return this.lane === lane && Math.abs(this.position - position) < 1;
  }

  /**
   * Check if obstacle is off the track
   * @param {number} trackLength - Length of the track
   * @returns {boolean} - True if obstacle is beyond track bounds
   */
  isOffTrack(trackLength) {
    return this.position < 0 || this.position > trackLength;
  }

  /**
   * Serialize obstacle to plain object
   * @returns {Object} - Serialized obstacle data
   */
  toObject() {
    return {
      lane: this.lane,
      position: this.position,
      type: this.type,
      symbol: this.symbol
    };
  }

  /**
   * Create a random obstacle
   * @param {number} position - Starting position (usually 0 for top of track)
   * @returns {Obstacle} - New obstacle instance
   */
  static createRandom(position = 0) {
    const lane = Math.floor(Math.random() * NUM_LANES);
    const type = Math.random() < 0.2 ? 'moving' : 'static'; // 20% chance of moving obstacle
    return new Obstacle(lane, position, type);
  }
}
